### PR TITLE
[Routing] Remove variadic constructor signature

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Just a little suggestion, based on today's "New in Symfony" blogpost.

Subjective argument: the variadic signature felt a bit weird, as argument 1 has a different meaning than 2+ (but they look very similar due to PHP's `::class` "magic constant")
Objective argument: variadic signature is harder with BC layers when we want to add another argument. Personally, I feel like we have to be careful about using variadic signatures in Symfony.

### CODE REVIEW

**Review by derrabus — APPROVED** [review]

**Review by stof — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Can you check if the build failure is related?

**Comment by wouterj** [comment]
Fixed fabbot, the CI test failures are caused by changes in PHP 8.2 upstream. You can see the same errors in e.g. https://github.com/symfony/symfony/runs/6146660539

Btw, let's mark PHP 8.2 tests as experimental again: https://github.com/symfony/symfony/pull/46160

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php [hunk 1]
```diff
@@ -22,12 +22,13 @@ final class EnumRequirement implements \Stringable
 
     /**
      * @template T of \BackedEnum
+     *
      * @param class-string<T> $enum
-     * @param T ...$cases
+     * @param list<T>         $cases
      */
-    public function __construct(string $enum, \BackedEnum ...$cases)
+    public function __construct(string $enum, array $cases = [])
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: `\get_class($case)` won't work when `$case` is not an object. We need to add a test covering this new possibility (`$cases` containing a scalar for example).

**Review by fancyweb — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: just change the code to use `get_debug_type` for the error message, and avoid accessing `$case->name` when it is not an enum.

Comment (wouterj) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: test added

**Comment by nicolas-grekas** [comment]
The benefit of using variadic was the type check enforced by PHP.
I'm really fine doing that but I'd suggest going one step further and allow only one argument: array|string $cases
When array, we should enforce that all cases have the same class.

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php [hunk 2]
```diff
@@ -22,18 +22,22 @@ final class EnumRequirement implements \Stringable
...
     {
-        if (!\is_subclass_of($enum, \BackedEnum::class, true)) {
+        if (!is_subclass_of($enum, \BackedEnum::class, true)) {
             throw new InvalidArgumentException(sprintf('"%s" is not a \BackedEnum class.', $enum));
         }
 
         foreach ($cases as $case) {
+            if (!$case instanceof \BackedEnum) {
+                throw new InvalidArgumentException(sprintf('Case "%s" is not a \BackedEnum class.', \get_debug_type($case)));
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Requirement/EnumRequirement.php]:
```suggestion
                throw new InvalidArgumentException(sprintf('Case must be a \BackedEnum instance, "%s" given.', \get_debug_type($case)));
```

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php [hunk 3]
```diff
@@ -22,18 +22,22 @@ final class EnumRequirement implements \Stringable
...
+            }
             if (!$case instanceof $enum) {
-                throw new InvalidArgumentException(sprintf('"%s::%s" is not a case of "%s".', \get_class($case), $case->name, $enum));
+                throw new InvalidArgumentException(sprintf('"%s::%s" is not a case of "%s".', \get_debug_type($case), $case->name, $enum));
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Requirement/EnumRequirement.php]:
```suggestion
                throw new InvalidArgumentException(sprintf('"%s::%s" is not a case of "%s".', \get_class($case), $case->name, $enum));
```
`get_class` is safe here.

**Review by fancyweb — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php (hunk 3 above)
Comment (wouterj) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: I have learned (from @nicolas-grekas IIRC) to always use `get_debug_type()` in favor of `get_class()`. It is bettering in dealing with e.g. anonymous classes.

Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: I agree with @wouterj: when a decision is non-conditional (using get_debug_type), everything is simpler.

Comment (fancyweb) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: Ok sorry :blush:

File: src/Symfony/Component/Routing/Tests/Requirement/EnumRequirementTest.php [hunk 4]
```diff
@@ -30,28 +30,36 @@ public function testNotABackedEnum()
         new EnumRequirement(TestUnitEnum::class);
     }
 
+    public function testCaseNotABackedEnum()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Case "string" is not a \BackedEnum class.');
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Tests/Requirement/EnumRequirementTest.php]:
```suggestion
        $this->expectExceptionMessage('Case must be a \BackedEnum instance, "string" given.');
```

**Review by fancyweb — APPROVED** [review]

**Comment by fancyweb** [comment]
> I'm really fine doing that but I'd suggest going one step further and allow only one argument: array|string $cases

:+1: or :-1:?

**Comment by wouterj** [comment]
:+1: for `string|array` - just updated the PR

**Comment by fancyweb** [comment]
It's actually `\BackedEnum|array`, I updated it.

**Comment by stof** [comment]
no, it is not `BackedEnum|array`. It it `class-string<BackedEnum>|array<BackedEnum>` , which is `string|array` as native typehint

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php [hunk 5]
```diff
@@ -22,18 +22,26 @@ final class EnumRequirement implements \Stringable
 
     /**
      * @template T of \BackedEnum
+     *
      * @param class-string<T> $enum
-     * @param T ...$cases
+     * @param T|list<T>       $cases
      */
-    public function __construct(string $enum, \BackedEnum ...$cases)
+    public function __construct(string $enum, \BackedEnum|array $cases = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: My proposal was to have only one argument ;)
(Please squash the commits btw)

Comment (fancyweb) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php]: @wouterj @stof do you agree? I misunderstood Nicolas' proposal.

**Review by nicolas-grekas — APPROVED** [review]
(I updated the PR to my proposal)

**Comment by fabpot** [comment]
Thank you @wouterj.

File: src/Symfony/Component/Routing/Requirement/EnumRequirement.php [hunk 6]
```diff
@@ -15,36 +15,42 @@
 
 final class EnumRequirement implements \Stringable
 {
-    /**
-     * @var string[]
-     */
-    private readonly array $values;
+    private string $requirement;
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Requirement/EnumRequirement.php:18]: @nicolas-grekas Why did you remove the readonly?

Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Requirement/EnumRequirement.php:18]: Not intentional! (But read-only on private is not useful either so 🤷‍♂️ 😆)
